plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp

```diff
@@ -440,32 +456,226 @@ class read_only {
+         auto& current_session = kv_database.get_kv_undo_stack()->top();
+         const auto value = current_session.read(full_key);
+         // check if we didn't actually find the key
+         if (!value) {
+            // only need to bother to do table search if we require it, so that we can report the correct error
+            if (require_table == row_requirements::required) {
+               const auto whole_table_prefix(backing_store::db_key_value_format::create_full_key_prefix(full_key, backing_store::db_key_value_format::end_of_prefix::pre_type));
+               const auto value = current_session.read(whole_table_prefix);
+               EOS_ASSERT(value, chain::contract_table_query_exception,
+                          "Missing code: ${code}, scope: ${scope}, table: ${table}",
+                          ("code",code.to_string())("scope",scope.to_string())("table",table.to_string()));
+            }
+            EOS_ASSERT(require_primary == row_requirements::optional, chain::contract_table_query_exception,
+                       "Missing row for primary_key: ${primary} in code: ${code}, scope: ${scope}, table: ${table}",
+                       ("primary", primary_key)("code",code.to_string())("scope",scope.to_string())
+                       ("table",table.to_string()));
+            return false;
+         }
+         f(chain::backing_store::primary_index_view::create(primary_key, value->data(), value->size()));
+         return true;
+      }
+   }
+
+   template<typename T, typename Function>
+   bool get_primary_key(name code, name scope, name table, uint64_t primary_key, row_requirements require_table,
+                        row_requirements require_primary, Function&& f) const {
+      auto ret = get_primary_key_internal(code, scope, table, primary_key, require_table, require_primary, [&f](const auto& obj) {
+         if( obj.value.size() >= sizeof(T) ) {
+            T t;
+            fc::datastream<const char *> ds(obj.value.data(), obj.value.size());
+            fc::raw::unpack(ds, t);
+
+            f(t);
+         }
+      });
+      return ret;
+   }
+
+   fc::variant get_primary_key(name code, name scope, name table, uint64_t primary_key, row_requirements require_table,
+                               row_requirements require_primary, const std::string_view& type, bool as_json = true) const;
+   fc::variant get_primary_key(name code, name scope, name table, uint64_t primary_key, row_requirements require_table,
+                               row_requirements require_primary, const std::string_view& type, const abi_serializer& abis,
+                               bool as_json = true) const;
+
+   auto get_primary_key_value(const std::string_view& type, const abi_serializer& abis, bool as_json = true) const {
+      return [&type,&abis,as_json,this](fc::variant& result_var, const auto& obj) {
```

Comment:
This assumes that `type` and `abis` and `this` live as long as the returned value. This is not always the case. For example, in `get_producers` a temporary string returned from `abs.get_table_type()` is passed in. After the call the temporary is destroyed and now you have undefined behavior.

Reply:
Thanks Kevin! I am fixing this now first and will do the others tomorrow.

Reply:
This should be fixed here not at the call site. It is an inherently bad pattern.

Reply:
Agreed, as I found another place in chain_plugin.hpp where uses the same way to call the function. Do you think simply capturing `type` by value by `return [type,&abis,as_json,this]` would work?

Reply:
`type` is a `string_view` so no, not unless you copy it into a string.

Reply:
This code changed, @linhuang-blockone is this already fixed?

Reply:
Yes, it is already fixed.